patroni/postgresql/citus.py

```diff
@@ -17,19 +17,299 @@
+            == set(self._node_hash(v, check_nodeid) for v in other)
+
+    def primary(self) -> Optional[PgDistNode]:
+        """Finds and returns :class:`PgDistNode` object that represents "primary"."""
+        return next(iter(v for v in self if v.is_primary()), None)
+
+    def get(self, value: PgDistNode) -> Optional[PgDistNode]:
+        """Performs a lookup of the actual value in a given set.
+
+        .. note::
+            It is necessary because :func:`__hash__` and :func:`__eq__` methods in :class:`PgDistNode`
+            are redefined and effectively they check only *host* and *port* attributes.
+
+        :param value: the key we search for.
+        :returns: the actual value from this :class:`set` object.
+        """
+        return next(iter(v for v in self if v == value), None)
+
+    def transition(self, old: 'PgDistGroup') -> Iterator[PgDistNode]:
+        """Compares this topology with the old one and yields transitions that transform the old to the new one.
+
+        .. note::
+            In addition to the yielding transactions this method fills up *nodeid*
+            attribute for nodes that are presented in the old and in the new topology.
+
+            There are a few simple rules/constraints that are imposed by Citus and must be followed:
+            - adding/removing nodes is only possible when metadata is synced to all registered "priorities".
+
+            - the "primary" row in "pg_dist_node" always keeps the nodeid (unless it is
+              removed, but it is not supported by Patroni).
+
+            - "nodename", "nodeport" must be unique across all rows in the "pg_dist_node".
+
+            - updating "broken" nodes always works and metadata is synced asynchnonously after commit.
+
+        Following these rules below is an example of the switchover between node1 (primary) and node2 (secondary).
+
+        .. code-block:: SQL
+
+            BEGIN;
+                SELECT citus_update_node(4, 'node1-demoted', 5432);
+                SELECT citus_update_node(5, 'node1', 5432);
+                SELECT citus_update_node(4, 'node2', 5432);
+            COMMIT;
+
+        :param old: the last know topology registered in "pg_dist_node" for a given *group*
+
+        :yields: :class:`PgDistNode` objects that must be updated/added/removed in "pg_dist_node".
+        """
+        self.failover = old.failover
+
+        new_primary = self.primary()
+        assert new_primary is not None
+        old_primary = old.primary()
+
+        gone_nodes = old - self - {old_primary}
+        added_nodes = self - old - {new_primary}
+
+        if not old_primary:
+            yield new_primary
+        elif old_primary == new_primary:
+            new_primary.nodeid = old_primary.nodeid
+            # Controlled switchover with pausing client connections.
+            # Achived by updating the primary row and putting hostname = '${host}-demoted' in a transaction.
+            if old_primary.role != new_primary.role:
```

Comment:
When we start switchover we will have `old_primary.role == 'primary'` and `new_primary.role == 'demoted'`.
After that there are two directions:
1. Some other node is promoted and switchover completes.
2. Despite all checks performed before demoting the primary there are no healthy replicas. In this case the old primary is promoted, and we will have to transition from `old_primary.role == 'demoted'` to `new_primary.role == 'primary'`.

I am not sure if it is possible to have `old_primary.role == 'demoted'` and `new_primary.role == 'demoted'`, but in case if it ever happens the current check will correctly handle it.